### PR TITLE
Extend support to GraniteMoeSharedForCausalLM architecture

### PR DESCRIPTION
**Pull request by kmehant**
Extend support to GraniteMoeSharedForCausalLM for all the fast moe (fast kernels, and EP) features and padding free.

### Summary of changes

- [x] extend support to GraniteMoeSharedForCausalLM
- [x] update a100_80gb_moe.csv |  bench for moe granite (moe and moe shared granite models)
- [x] update a100_80gb_moe.csv | bench for mixtral
- [x] update [requirements_moe.txt](https://github.com/foundation-model-stack/fms-acceleration/blob/main/scripts/benchmarks/refs/requirements_moe.txt)
- [x] regression test for the above runs.

### Final reg plots that includes granite moe and mixtral.

| memory increased             |  loss holds parity | throughput increased |
:-------------------------:|:-------------------------:|:-------------------------:
compare-mem_nvidia_mem_reserved  |  compare-train_loss | compare-train_tokens_per_second

### Outliers
Screenshot 2025-02-21 at 12 48 31 PM

### OOM experiments

**Mixtral OOM due to extra memory consumption**
epoch | framework_config | gradient_accumulation_steps | mem_nvidia_mem_reserved | model_name_or_path | num_gpus | per_device_train_batch_size | torch_dtype | train_loss | train_runtime | train_samples_per_second | train_steps_per_second | train_tokens_per_second
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
  | none | 16 | 78783.5 | mistralai/Mixtral-8x7B-Instruct-v0.1 | 8 | 1 | bfloat16 |   |   |   |   |  

Also note:
- `ibm-research/moe-7b-1b-active-shared-experts` has number of experts that are not divisible by 4, so `ep4` does not apply

Reference issue on regression - https://github.com/foundation-model-stack/fms-acceleration/issues/128

### CODE REVIEW

File: plugins/fused-ops-and-kernels/src/fms_acceleration_foak/models/granitemoeshared.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+                partial(
+                    build_lora_fused_ops,
+                    submodule_names=["o_proj"],
+                    fused_op=KEY_O,
+                    base_type=base_type,
+                ),
+                logic="APPEND",
+            ),
+        ),
+        ModelPatcherRule(
```
Comment (fabianlim) [inline plugins/fused-ops-and-kernels/src/fms_acceleration_foak/models/granitemoeshared.py:96]: refer to how [granite](https://github.com/foundation-model-stack/fms-acceleration/blob/23ea8ec64a72a866d5ea786f63b41db12f9bec72/plugins/fused-ops-and-kernels/src/fms_acceleration_foak/models/granite.py#L129) does it

File: scripts/benchmarks/scenarios-moe.yaml [hunk 2]
```diff
@@ -59,6 +59,47 @@ scenarios:
...
+        arguments:
+            learning_rate: 5e-5
+            torch_dtype: bfloat16
+            gradient_accumulation_steps: null
+            per_device_train_batch_size: 8
+            logging_steps: 1
+            packing: False
+            adam_epsilon: 1e-8
+            model_name_or_path: 
+                - 'ibm-research/moe-7b-1b-active-shared-experts'
```
Comment (fabianlim) [inline scripts/benchmarks/scenarios-moe.yaml:83]: you shouldnt copy and paste this, you should just add to `model_name_or_path` in the existing scenarios

```
model_name_or_path:
  - 'ibm-granite/granite-3.0-3b-a800m-instruct'
  - 'ibm-research/moe-7b-1b-active-shared-experts'
 ```

File: scripts/benchmarks/scenarios-moe.yaml [hunk 3]
```diff
@@ -59,6 +59,47 @@ scenarios:
...
+
+    -   name: accelerated-moe-full-granite-moe-shared-small
+        framework_config: 
+            - # without acceleration
+            - moe-scattermoe-granite-ep4
+            - moe-scattermoe-granite-ep4-padding-free
+            - moe-scattermoe-granite-ep4-padding-free-foak
+        arguments:
+            learning_rate: 5e-5
+            torch_dtype: bfloat16
```
Comment (fabianlim) [inline scripts/benchmarks/scenarios-moe.yaml:93]: same here.. dont copy and paste.. if its all the same arguments there is no need

I dont understand why you need this different bench

File: scripts/benchmarks/benchmark.py [hunk 4]
```diff
@@ -319,10 +319,11 @@ def build_args_from_products(products: List[Dict], defaults: Dict):
                     ]
                 )
             elif grad_accum is None and pdtbs is not None:
+                gas = effective_batch_size // num_gpus // pdtbs
                 argument_list.extend(
                     [
                         "--gradient_accumulation_steps",
-                        str(effective_batch_size // num_gpus // pdtbs),
+                        str(1 if gas == 0 else gas),
```
Comment (fabianlim) [inline scripts/benchmarks/benchmark.py:326]: ok this worjks, but dont understand why you need it, because your benches use the same parameters as the existing ones, and we dont run into this issue

**Review by fabianlim — CHANGES_REQUESTED** [review]

File: scripts/benchmarks/scenarios-moe.yaml (hunk 3 above)
Comment (kmehant) [reply scripts/benchmarks/scenarios-moe.yaml:93]: Wanted to run only for moesharedmodel so had to copy paste. Is there a way I can subselect a model along with scenario with `scenariofilter`? Apologies if I have missed that.

File: scripts/benchmarks/benchmark.py (hunk 4 above)
Comment (kmehant) [reply scripts/benchmarks/benchmark.py:326]: Yes, @fabianlim. I am not sure how your benches passed through this value error - https://github.com/foundation-model-stack/fms-hf-tuning/blob/fb3ace8397223932e176de604703c54a14e1ebf0/tuning/sft_trainer.py#L136-L139

Comment (kmehant) [reply scripts/benchmarks/benchmark.py:326]: One thing I noticed is the experiments continue to happen silently even when some of them are failed and also the benchmark report gets generated irrespectively,

**Review by willmj — COMMENTED** [review]
Once Fabian's comments have been addressed this looks good to me except for minor nit, would be interested to see benchmarks as well if available.

File: scripts/benchmarks/scenarios-moe.yaml (hunk 3 above)
Comment (fabianlim) [reply scripts/benchmarks/scenarios-moe.yaml]: sorry  you cant. If you want to do ad hoc testing then just uncomment the other models you dont want to test. For the official bench we need to update all models, this is because we only version 1 set of requirements for reproducibilty, and we cant have partial benches running, otherwise there will be inconsistency

**Comment by fabianlim** [comment]
@kmehant yes code changes looks good, what is missing now is the updating of the benches. So you need to run the full bench for [scenarios-moe.yaml](https://github.com/foundation-model-stack/fms-acceleration/blob/main/scripts/benchmarks/scenarios-moe.yaml) now in order to update [a100_80gb_moe.csv](https://github.com/foundation-model-stack/fms-acceleration/blob/main/scripts/benchmarks/refs/a100_80gb_moe.csv) and [requirements_moe.txt](https://github.com/foundation-model-stack/fms-acceleration/blob/main/scripts/benchmarks/refs/requirements_moe.txt)
- Please run the bench using the documented [instructions](https://github.com/foundation-model-stack/fms-acceleration/tree/main/plugins/accelerated-moe#running-benchmarks), running the actual bench in adhoc ways cause inconsistencies.
- There are two suites `accelerated-moe-scatter` and `accelerated-moe-scatter-mixtral`, their commands are seperate, and then the results are stictched together. If this is too troublesome, then you can split them. Take out the `mixtral` benches and put them in seperate requirements and CSV file (maybe call them `moe_mixtral.csv`, `requirements_moe_mixtral.csv`, etc.. 
- Then run one full bench for `accelerated-moe-scatter` for both models and update accordingly

**Comment by kmehant** [comment]
bench for moe granite (moe and moe shared granite models)
bench for mixtral

**Comment by fabianlim** [comment]
@willmj and @anhuong can either of you help to guide @kmehant how to produce the regression plots?

**Comment by kmehant** [comment]
https://github.com/foundation-model-stack/fms-acceleration/pull/126#issuecomment-2670670238

Got it @fabianlim do you have those scripts handy?

**Comment by kmehant** [comment]
### Use of memory increased
compare-mem_nvidia_mem_reserved

### Train loss parity exists
compare-train_loss

### Throughput increased in the new benchmark 
compare-train_tokens_per_second

**Comment by kmehant** [comment]
Outliers
Screenshot 2025-02-20 at 2 31 22 PM

**Comment by kmehant** [comment]
@fabianlim apologies for the confusion. I was sharing intermediate results and regression. I am tracking it [here](https://github.com/foundation-model-stack/fms-acceleration/pull/126#issuecomment-2670653771) and mixtral is currently running and will update the regression and results once done. Thank you.

**Comment by kmehant** [comment]
Final reg plots that includes mixtral.


| memory increased             |  loss holds parity | throughput increased |
:-------------------------:|:-------------------------:|:-------------------------:
compare-mem_nvidia_mem_reserved  |  compare-train_loss | compare-train_tokens_per_second

File: scripts/benchmarks/refs/requirements_moe.txt [hunk 5]
```diff
@@ -1,44 +1,42 @@
...
-fonttools==4.54.1
+einops==0.8.1
+filelock==3.17.0
+flash_attn==2.7.4.post1
+-e git+https://github.com/kmehant/fms-acceleration.git@d0560e0c652dc2f72fdad07a2fb7af8fe67bbf08#egg=fms_acceleration&subdirectory=plugins/framework
+-e git+https://github.com/kmehant/fms-acceleration.git@d0560e0c652dc2f72fdad07a2fb7af8fe67bbf08#egg=fms_acceleration_aadp&subdirectory=plugins/attention-and-distributed-packing
+-e git+https://github.com/kmehant/fms-acceleration.git@d0560e0c652dc2f72fdad07a2fb7af8fe67bbf08#egg=fms_acceleration_foak&subdirectory=plugins/fused-ops-and-kernels
+-e git+https://github.com/kmehant/fms-acceleration.git@d0560e0c652dc2f72fdad07a2fb7af8fe67bbf08#egg=fms_acceleration_moe&subdirectory=plugins/accelerated-moe
+-e git+https://github.com/kmehant/fms-acceleration.git@d0560e0c652dc2f72fdad07a2fb7af8fe67bbf08#egg=fms_acceleration_peft&subdirectory=plugins/accelerated-peft
+fms-hf-tuning @ git+https://github.com/foundation-model-stack/fms-hf-tuning.git@fdc7527510692ada03e4303df1549cebc5139b31
```
Comment (kmehant) [inline scripts/benchmarks/refs/requirements_moe.txt:21]: I did a pip freeze to update this, however, this would show my forks of fms-acceleration. Should we keep it as is?

Comment (fabianlim) [reply scripts/benchmarks/refs/requirements_moe.txt:21]: no pls delete

File: tox.ini [hunk 6]
```diff
@@ -42,7 +42,7 @@ commands =
     python -m fms_acceleration.cli install -e {toxinidir}/plugins/accelerated-moe
 
     # install the flash attn at the last 
-    pip install flash-attn
+    pip install flash-attn --no-build-isolation
```
Comment (fabianlim) [inline tox.ini:45]: does this really need to change

Comment (kmehant) [reply tox.ini:45]: I had this `torch module not found issue` with tox. Had to do this.

**Review by fabianlim — APPROVED** [review]
@kmehant can you put your final regression plots into the main description. Otherwise i think it looks good. Lets have @willmj take one look at it before merge

File: tox.ini (hunk 6 above)
Comment (fabianlim) [reply tox.ini:45]: im abit confused because i believe torch is installed with `fms_hf_tuing` and our tox file installs that

Comment (fabianlim) [reply tox.ini:45]: how about this.. maybe can you leave out this commit, and make an issue saying you want to add it in

File: scripts/benchmarks/refs/requirements_moe.txt (hunk 5 above)
Comment (kmehant) [reply scripts/benchmarks/refs/requirements_moe.txt]: updated the req file. Thanks

**Review by willmj — APPROVED** [review]
LGTM

**Comment by kmehant** [comment]
@fabianlim @willmj requesting merge.

**Comment by fabianlim** [comment]
@kmehant sorry i just noticed you only have 2 mixtral entries, but we used to have 3. can you check this 
image

**Comment by kmehant** [comment]
https://github.com/foundation-model-stack/fms-acceleration/pull/126#issuecomment-2675013477

The FSDP variant which was previously part got OOMed @fabianlim

**Comment by fabianlim** [comment]
All the OOMs need to be recorded down, can you do the same in the main commeng @kmehant , and pls open an issue and link it here for tracking

**Comment by kmehant** [comment]
https://github.com/foundation-model-stack/fms-acceleration/pull/126#issuecomment-2675819101

@fabianlim completed!
